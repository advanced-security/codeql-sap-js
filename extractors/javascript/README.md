# Extension to the JavaScript extractor to support CDS compilation

This directory contains a `pre-finalize.sh` script that can be dropped into the `tools` directory
of a JavaScript extractor to enable the automatic compilation and extraction of SAP CAP CDS files
during the `codeql database create` process.

The script requires that the `cds` extractor is available when `codeql database init` is called. This can either be through providing the extractors parent directory using the `--search-path` bundle, or by including the extractor (and this pre-finalize script) in a custom bundle created by the [CodeQL Development Toolkit (qlt)](https://github.com/advanced-security/codeql-development-toolkit).