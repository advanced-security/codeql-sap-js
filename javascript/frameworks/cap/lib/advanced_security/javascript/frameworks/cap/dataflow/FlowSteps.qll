/**
 * Additional flow steps to be registered to `DataFlow::SharedFlowStep`.
 */