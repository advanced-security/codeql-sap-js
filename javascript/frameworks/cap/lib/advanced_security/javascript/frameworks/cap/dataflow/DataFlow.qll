/**
 * Security-related `DataFlow::Node`s or relations between two `DataFlow::Node`s.
 */

import javascript
import advanced_security.javascript.frameworks.cap.CDS

/**
 * Methods that parse source strings into a CQL expression.
 */
class ParseSink extends DataFlow::Node {
  ParseSink() {
    exists(CdsFacade cds |
      this = cds.getMember("parse").getMember(["expr", "ref", "xpr"]).getACall().getAnArgument()
    )
  }
}

/**
 * A communication happening between `cds.Service`s. This includes:
 * 1. Ones based on REST-style API, based on `cds.Service.send`,
 * 2. Ones based on query-style API, based on `cds.Services.run`, and
 * 3. Ones based on emitting and subscribing to asynchronous event messages.
 */
abstract class InterServiceCommunication extends MethodCallNode {
  InterServiceCommunication() { this.getReceiver() instanceof ServiceInstance }

  /* TODO: Generalize UserApplicationService to include built-in services such as log and db */
  /**
   * The service that sends the request.
   */
  UserDefinedApplicationService sender;
  /**
   * The service that receives the request and handles it.
   */
  UserDefinedApplicationService recipient;

  UserDefinedApplicationService getSender() { result = sender }

  UserDefinedApplicationService getReceipient() { result = recipient }
}

class RestStyleCommunication extends InterServiceCommunication {
  RestStyleCommunication() {
    exists(HandlerRegistration registration, SrvSend srvSend |
      sender = registration.getReceiver().(ServiceInstance).getDefinition() and
      recipient = srvSend.getReceiver().(ServiceInstance).getDefinition()
    )
  }
}

class CrudStyleCommunication extends InterServiceCommunication {
  CrudStyleCommunication() {
    exists(HandlerRegistration registration, SrvRun srvRun |
      sender = registration.getReceiver().(ServiceInstance).getDefinition() and
      recipient = srvRun.getReceiver().(ServiceInstance).getDefinition()
    )
  }
}

class AsyncStyleCommunication extends InterServiceCommunication {
  AsyncStyleCommunication() {
    exists(
      HandlerRegistration emittingRegistration, HandlerRegistration orchestratingRegistration,
      SrvEmit srvEmit, InterServiceCommunicationMethodCall methodCallOnReceiver
    |
      emittingRegistration != orchestratingRegistration and
      /* The service that emits the event and the service that registers the handler are the same; it's the sender. */
      this = orchestratingRegistration and
      sender = emittingRegistration.getReceiver().(ServiceInstance).getDefinition() and
      srvEmit.asExpr().getEnclosingFunction+() = sender.getInitFunction().asExpr() and
      /* 1. match by their event name. */
      srvEmit.getEmittedEvent() = orchestratingRegistration.getAnEventName() and
      /* 2. match by their service name in cds.connect().to(). */
      srvEmit.getEmitter().getDefinition().getManifestName() =
        orchestratingRegistration.getReceiver().(ServiceInstanceFromCdsConnectTo).getServiceName() and
      recipient = methodCallOnReceiver.getReceiver().(ServiceInstance).getDefinition() and
      methodCallOnReceiver.getEnclosingFunction() = orchestratingRegistration.getHandler().asExpr()
    )
  }
}
