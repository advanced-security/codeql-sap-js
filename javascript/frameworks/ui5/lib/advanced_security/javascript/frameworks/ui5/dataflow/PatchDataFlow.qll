/**
 * This file patches an incompatibility introduced into the standard data flow library between
 * class DataFlow::Configurations and `summmaryModels` added in models-as-data files, and likely
 * introduced in this PR: https://github.com/github/codeql/pull/19445/files.
 */

import javascript
import semmle.javascript.dataflow.internal.FlowSummaryPrivate
private import semmle.javascript.frameworks.data.internal.ApiGraphModels as Shared

/**
 * Holds if `path` is an input or output spec for a summary with the given `base` node.
 */
pragma[nomagic]
private predicate relevantInputOutputPath(API::InvokeNode base, AccessPath inputOrOutput) {
  exists(string type, string input, string output, string path |
    ModelOutput::resolvedSummaryBase(type, path, base) and
    ModelOutput::relevantSummaryModel(type, path, input, output, _, _) and
    inputOrOutput = [input, output]
  )
}

/**
 * Gets the API node for the first `n` tokens of the given input/output path, evaluated relative to `baseNode`.
 */
private API::Node getNodeFromInputOutputPath(API::InvokeNode baseNode, AccessPath path, int n) {
  relevantInputOutputPath(baseNode, path) and
  (
    n = 1 and
    result = Shared::getSuccessorFromInvoke(baseNode, path.getToken(0))
    or
    result =
      Shared::getSuccessorFromNode(getNodeFromInputOutputPath(baseNode, path, n - 1),
        path.getToken(n - 1))
  )
}

/**
 * Gets the API node for the given input/output path, evaluated relative to `baseNode`.
 */
private API::Node getNodeFromInputOutputPath(API::InvokeNode baseNode, AccessPath path) {
  result = getNodeFromInputOutputPath(baseNode, path, path.getNumToken())
}

private predicate summaryStep(API::Node pred, API::Node succ, string kind) {
  exists(string type, string path, API::InvokeNode base, AccessPath input, AccessPath output |
    ModelOutput::relevantSummaryModel(type, path, input, output, kind, _) and
    ModelOutput::resolvedSummaryBase(type, path, base) and
    pred = getNodeFromInputOutputPath(base, input) and
    succ = getNodeFromInputOutputPath(base, output)
  )
}

/**
 * Like `ModelOutput::summaryStep` but with API nodes mapped to data-flow nodes.
 */
private predicate summaryStepNodes(DataFlow::Node pred, DataFlow::Node succ, string kind) {
  exists(API::Node predNode, API::Node succNode |
    summaryStep(predNode, succNode, kind) and
    pred = predNode.asSink() and
    succ = succNode.asSource()
  )
}

/** Data flow steps induced by summary models of kind `value`. */
private class DataFlowStepFromSummary extends DataFlow::SharedFlowStep {
  override predicate step(DataFlow::Node pred, DataFlow::Node succ) {
    summaryStepNodes(pred, succ, "value")
  }
}

/** Taint steps induced by summary models of kind `taint`. */
private class TaintStepFromSummary extends TaintTracking::SharedTaintStep {
  override predicate step(DataFlow::Node pred, DataFlow::Node succ) {
    summaryStepNodes(pred, succ, "taint")
  }
}
