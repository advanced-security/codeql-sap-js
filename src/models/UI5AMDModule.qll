/**
 * Provides classes for working with
 * [Asynchronous Module Definitions](https://github.com/amdjs/amdjs-api/wiki/AMD).
 */

import javascript
private import semmle.javascript.internal.CachedStages
private import Expressions.ExprHasNoEffect

/**
 * A IU5 `declare`/`require` call using jQuery.
 *
 * Example:
 *
 * ```
 * âˆš("codeql-sap-js.control.xss");
 *
 * The first argument is a dependency.
 * ```
 */
abstract class SapJQueryModuleDefinition extends CallExpr {
  SapJQueryModuleDefinition() {
    inVoidContext(this) and
    this.getReceiver().(PropAccess).getQualifiedName() = "jQuery.sap"
  }
}

/**
 * `jQuery.sap.declare` call
 */
class SapJQueryDefine extends SapJQueryModuleDefinition {
  SapJQueryDefine() { this.getCalleeName() = "declare" }
}

/**
 * `jQuery.sap.require` call
 */
class SapJQueryRequire extends SapJQueryModuleDefinition {
  SapJQueryRequire() { this.getCalleeName() = "require" }
}

/**
 * An AMD `define`/`require` call.
 *
 * Example:
 *
 * ```
 * sap.ui.define(['fs', 'express'], function(fs, express) {
 *   ...
 * });
 * ```
 *
 * The first argument is an (optional) array of dependencies,
 * the second a factory method or object.
 *
 * We also recognize the three-argument form `define('m', ['fs', 'express'], ...)`
 * where the first argument is the module name, the second argument an
 * array of dependencies, and the third argument a factory method or object.
 */
abstract class SapAmdModuleDefinition extends CallExpr {
  SapAmdModuleDefinition() {
    inVoidContext(this) and
    exists(int n | n = this.getNumArgument() |
      n = 1
      or
      n = 2 and this.getArgument(0) instanceof ArrayExpr
      or
      n = 3 and
      this.getArgument(0) instanceof ConstantString and
      this.getArgument(1) instanceof ArrayExpr
    )
  }

  /** Gets the array of module dependencies, if any. */
  ArrayExpr getDependencies() {
    result = this.getArgument(0) or
    result = this.getArgument(1)
  }

  /** Gets the `i`th dependency of this module definition. */
  PathExpr getDependency(int i) { result = this.getDependencies().getElement(i) }

  /** Gets a dependency of this module definition. */
  PathExpr getADependency() {
    result = this.getDependency(_) or
    result = this.getARequireCall().getAnArgument()
  }

  /**
   * Gets a data flow node containing the factory value of this module definition.
   */
  pragma[nomagic]
  DataFlow::SourceNode getFactoryNode() {
    result = this.getFactoryNodeInternal() and
    result instanceof DataFlow::ValueNode
  }

  private DataFlow::Node getFactoryNodeInternal() {
    // To avoid recursion, this should not depend on `SourceNode`.
    result = DataFlow::valueNode(this.getLastArgument()) or
    result = this.getFactoryNodeInternal().getAPredecessor()
  }

  /** Gets the expression defining this module. */
  Expr getModuleExpr() {
    exists(DataFlow::Node f | f = this.getFactoryNode() |
      if f instanceof DataFlow::FunctionNode
      then
        exists(ReturnStmt ret | ret.getContainer() = f.(DataFlow::FunctionNode).getAstNode() |
          result = ret.getExpr()
        )
      else result = f.asExpr()
    )
  }

  /** Gets a source node whose value becomes the definition of this module. */
  DataFlow::SourceNode getAModuleSource() { result.flowsToExpr(this.getModuleExpr()) }

  /**
   * Holds if `p` is the parameter corresponding to dependency `dep`.
   */
  predicate dependencyParameter(PathExpr dep, Parameter p) {
    exists(int i |
      dep = this.getDependency(i) and
      p = this.getFactoryParameter(i)
    )
  }

  /**
   * Gets the parameter corresponding to dependency `name`.
   *
   * For instance, in the module definition
   *
   * ```
   * define(['dep1', 'dep2'], function(pdep1, pdep2) { ... })
   * ```
   *
   * parameters `pdep1` and `pdep2` correspond to dependencies
   * `dep1` and `dep2`.
   */
  Parameter getDependencyParameter(string name) {
    exists(PathExpr dep |
      this.dependencyParameter(dep, result) and
      dep.getValue() = name
    )
  }

  /**
   * Gets the `i`th parameter of the factory function of this module.
   */
  private Parameter getFactoryParameter(int i) {
    this.getFactoryNodeInternal().asExpr().(Function).getParameter(i) = result
  }

  /**
   * Gets the parameter corresponding to the pseudo-dependency `require`.
   */
  Parameter getRequireParameter() {
    result = this.getDependencyParameter("require")
    or
    // if no dependencies are listed, the first parameter is assumed to be `require`
    not exists(this.getDependencies()) and result = this.getFactoryParameter(0)
  }

  pragma[noinline]
  private Variable getRequireVariable() { result = this.getRequireParameter().getVariable() }

  /**
   * Gets the parameter corresponding to the pseudo-dependency `exports`.
   */
  Parameter getExportsParameter() {
    result = this.getDependencyParameter("exports")
    or
    // if no dependencies are listed, the second parameter is assumed to be `exports`
    not exists(this.getDependencies()) and result = this.getFactoryParameter(1)
  }

  /**
   * Gets the parameter corresponding to the pseudo-dependency `module`.
   */
  Parameter getModuleParameter() {
    result = this.getDependencyParameter("module")
    or
    // if no dependencies are listed, the third parameter is assumed to be `module`
    not exists(this.getDependencies()) and result = this.getFactoryParameter(2)
  }

  /**
   * Gets an abstract value representing one or more values that may flow
   * into this module's `module.exports` property.
   */
  DefiniteAbstractValue getAModuleExportsValue() {
    result = [this.getAnImplicitExportsValue(), this.getAnExplicitExportsValue()]
  }

  pragma[noinline, nomagic]
  private AbstractValue getAnImplicitExportsValue() {
    // implicit exports: anything that is returned from the factory function
    result = this.getModuleExpr().analyze().getAValue()
  }

  pragma[noinline]
  private AbstractValue getAnExplicitExportsValue() {
    // explicit exports: anything assigned to `module.exports`
    exists(AbstractProperty moduleExports, SapAmdModule m |
      this = m.getDefine() and
      moduleExports.getBase().(AbstractModuleObject).getModule() = m and
      moduleExports.getPropertyName() = "exports"
    |
      result = moduleExports.getAValue()
    )
  }

  /**
   * Gets a call to `require` inside this module.
   */
  CallExpr getARequireCall() {
    result.getCallee().getUnderlyingValue() = this.getRequireVariable().getAnAccess()
  }
}

/**
 * `sap.ui.define` call
 */
class SapUiDefine extends SapAmdModuleDefinition {
  SapUiDefine() {
    this.getReceiver().(PropAccess).getQualifiedName() = "sap.ui" and
    this.getCalleeName() = "define"
  }
}

/**
 * `sap.ui.require` call
 */
class SapUiRequire extends SapAmdModuleDefinition {
  SapUiRequire() {
    this.getReceiver().(PropAccess).getQualifiedName() = "sap.ui" and
    this.getCalleeName() = "require"
  }
}

/** An UI5 dependency, considered as a path expression. */
private class SapAmdDependencyPath extends PathExprCandidate {
  SapAmdDependencyPath() {
    exists(SapAmdModuleDefinition amd |
      this = amd.getDependencies().getAnElement() or
      this = amd.getARequireCall().getAnArgument()
    )
  }
}

/** A constant path element appearing in an AMD dependency expression. */
private class ConstantAmdDependencyPathElement extends PathExpr, ConstantString {
  ConstantAmdDependencyPathElement() { this = any(SapAmdDependencyPath amd).getAPart() }

  override string getValue() { result = this.getStringValue() }
}

/**
 * Holds if `nd` is nested inside an AMD module definition.
 */
private predicate inAmdModuleDefinition(AstNode nd) {
  nd.getParent() instanceof SapAmdModuleDefinition
  or
  inAmdModuleDefinition(nd.getParent())
}

/**
 * Holds if `def` is an AMD module definition in `tl` which is not
 * nested inside another module definition.
 */
private predicate amdModuleTopLevel(SapAmdModuleDefinition def, TopLevel tl) {
  def.getTopLevel() = tl and
  not inAmdModuleDefinition(def)
}

/**
 * An AMD dependency, viewed as an import.
 */
private class SapAmdDependencyImport extends Import {
  SapAmdDependencyImport() { this = any(SapAmdModuleDefinition def).getADependency() }

  override Module getEnclosingModule() { this = result.(SapAmdModule).getDefine().getADependency() }

  override PathExpr getImportedPath() { result = this }

  /**
   * Gets a file that looks like it might be the target of this import.
   *
   * Specifically, we look for files whose absolute path ends with the imported path, possibly
   * adding well-known JavaScript file extensions like `.js`.
   */
  private File guessTarget() {
    exists(PathString imported, string abspath, string dirname, string basename |
      this.targetCandidate(result, abspath, imported, dirname, basename)
    |
      abspath.regexpMatch(".*/\\Q" + imported + "\\E")
      or
      exists(Folder dir |
        // `dir` ends with the dirname of the imported path
        dir.getAbsolutePath().regexpMatch(".*/\\Q" + dirname + "\\E") or
        dirname = ""
      |
        result = dir.getJavaScriptFile(basename)
      )
    )
  }

  /**
   * Holds if `f` is a file whose stem (that is, basename without extension) matches the imported path.
   *
   * Additionally, `abspath` is bound to the absolute path of `f`, `imported` to the imported path, and
   * `dirname` and `basename` to the dirname and basename (respectively) of `imported`.
   */
  private predicate targetCandidate(
    File f, string abspath, PathString imported, string dirname, string basename
  ) {
    imported = this.getImportedPath().getValue() and
    f.getStem() = imported.getStem() and
    f.getAbsolutePath() = abspath and
    dirname = imported.getDirName() and
    basename = imported.getBaseName()
  }

  /**
   * Gets the module whose absolute path matches this import, if there is only a single such module.
   */
  private Module resolveByAbsolutePath() {
    result.getFile() = unique(File file | file = this.guessTarget())
  }

  override Module getImportedModule() {
    result = super.getImportedModule()
    or
    not exists(super.getImportedModule()) and
    result = this.resolveByAbsolutePath()
  }

  override DataFlow::Node getImportedModuleNode() {
    exists(Parameter param |
      any(SapAmdModuleDefinition def).dependencyParameter(this, param) and
      result = DataFlow::parameterNode(param)
    )
  }
}

/**
 * An AMD-style module.
 *
 * Example:
 *
 * ```
 * define(['fs', 'express'], function(fs, express) {
 *   ...
 * });
 * ```
 */
class SapAmdModule extends Module {
  cached
  SapAmdModule() {
    Stages::DataFlowStage::ref() and
    exists(unique(SapAmdModuleDefinition def | amdModuleTopLevel(def, this)))
  }

  /** Gets the definition of this module. */
  SapAmdModuleDefinition getDefine() { amdModuleTopLevel(result, this) }

  override DataFlow::Node getAnExportedValue(string name) {
    exists(DataFlow::PropWrite pwn | result = pwn.getRhs() |
      pwn.getBase().analyze().getAValue() = this.getDefine().getAModuleExportsValue() and
      name = pwn.getPropertyName()
    )
  }

  override DataFlow::Node getABulkExportedNode() {
    // Assigned to `module.exports` via the factory's `module` parameter
    exists(AbstractModuleObject m, DataFlow::PropWrite write |
      m.getModule() = this and
      write.getPropertyName() = "exports" and
      write.getBase().analyze().getAValue() = m and
      result = write.getRhs()
    )
    or
    // Returned from factory function
    result = this.getDefine().getModuleExpr().flow()
  }
}
