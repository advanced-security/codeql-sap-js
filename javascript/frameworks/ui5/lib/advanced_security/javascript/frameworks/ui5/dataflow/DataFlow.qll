import javascript
import semmle.javascript.dataflow.DataFlow as StdLibDataFlow
import advanced_security.javascript.frameworks.ui5.UI5
import advanced_security.javascript.frameworks.ui5.UI5View
import advanced_security.javascript.frameworks.ui5.RemoteFlowSources
import advanced_security.javascript.frameworks.ui5.dataflow.FlowSteps
private import PatchDataFlow

/**
 * A statically visible part of a local model's content that has a binding path referring to it in a
 * control declaration acting as an HTML injection sink.
 *
 * e.g.1. Given a JSON model `oModel` declared in a controller handler and an HTML injection sink
 * `SomeSinkControl` as:
 * ```javascript
 * let oModel = new JSONModel({ y: null });
 * ```
 * and
 * ```xml
 * <SomeSinkControl x="{/y}"/>
 * ```
 * The content `y: null` of `oModel` is recognized as an instance of this class.
 *
 * e.g.2. Given a JSON model `oModel` which gains its content from a JSON file and an HTML injection
 * sink `SomeSinkControl` as:
 * ```javascript
 * let oModel = new JSONModel("controller/contents.json");
 * ```
 * and
 * ```xml
 * <SomeSinkControl x="{/y}"/>
 * ```
 * where `controller/contents.json` contains
 * ```json
 * { "y": null }
 * ```
 * The content `y: null` of `oModel` is recognized as an instance of this class.
 */
class LocalModelContentBoundBidirectionallyToHtmlISinkControl extends DomBasedXss::Sink {
  UI5BindingPath bindingPath;
  UI5Control controlDeclaration;

  LocalModelContentBoundBidirectionallyToHtmlISinkControl() {
    exists(UI5InternalModel internalModel |
      this = bindingPath.getNode() and
      (
        this instanceof PropWrite and
        internalModel.getArgument(0).getALocalSource().asExpr() =
          this.(PropWrite).getPropertyNameExpr().getParent+()
        or
        this.asExpr() instanceof StringLiteral and
        internalModel.asExpr() = this.asExpr().getParent()
      ) and
      any(UI5View view).getAnHtmlISink() = bindingPath and
      internalModel.(JsonModel).isTwoWayBinding() and
      controlDeclaration = bindingPath.getControlDeclaration()
    )
  }

  UI5BindingPath getBindingPath() { result = bindingPath }

  UI5Control getControlDeclaration() { result = controlDeclaration }
}

module UI5PathGraph<PathNodeSig ConfigPathNode, PathGraphSig<ConfigPathNode> ConfigPathGraph> {
  private newtype TNode =
    TUI5BindingPathNode(UI5BindingPath path) or
    TDataFlowNode(DataFlow::Node node)

  class UI5PathNode extends TNode {
    DataFlow::Node asDataFlowNode() { this = TDataFlowNode(result) }

    UI5BindingPath asUI5BindingPathNode() { this = TUI5BindingPathNode(result) }

    predicate isDataFlowNode() { exists(this.asDataFlowNode()) }

    predicate isUI5BindingPathNode() { exists(this.asUI5BindingPathNode()) }

    string toString() {
      result = this.asDataFlowNode().toString()
      or
      result = this.asUI5BindingPathNode().toString()
    }

    File getFile() {
      result = this.asDataFlowNode().getFile()
      or
      result = this.asUI5BindingPathNode().getView()
    }

    predicate hasLocationInfo(
      string filepath, int startline, int startcolumn, int endline, int endcolumn
    ) {
      this.asDataFlowNode().hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
      or
      this.asUI5BindingPathNode()
          .getLocation()
          .hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
    }

    ConfigPathNode getPathNode() { result.getNode() = this.asDataFlowNode() }

    UI5PathNode getAPrimarySource() {
      if this.asDataFlowNode() instanceof LocalModelContentBoundBidirectionallyToSourceControl
      then
        result.asUI5BindingPathNode() =
          this.asDataFlowNode()
              .(LocalModelContentBoundBidirectionallyToSourceControl)
              .getBindingPath()
      else result = this
    }

    UI5PathNode getAPrimaryHtmlISink() {
      if
        this.asDataFlowNode() instanceof LocalModelContentBoundBidirectionallyToHtmlISinkControl or
        this.asDataFlowNode() instanceof UI5ExternalModel // TODO: Narrow it down to ExternalModelBoundToHtmlISinkControl
      then
        result.asUI5BindingPathNode() =
          this.asDataFlowNode()
              .(LocalModelContentBoundBidirectionallyToHtmlISinkControl)
              .getBindingPath()
        or
        /* Look for the control with the binding path associated with the model */
        result.asUI5BindingPathNode().getModel().(UI5ExternalModel) = this.asDataFlowNode() and
        result.asUI5BindingPathNode() = any(UI5View view).getAnHtmlISink()
      else result = this
    }
  }

  query predicate nodes(UI5PathNode ui5PathNode) {
    exists(ui5PathNode.asUI5BindingPathNode())
    or
    exists(ConfigPathNode pathNode |
      pathNode.getNode() = ui5PathNode.asDataFlowNode() and
      ConfigPathGraph::nodes(pathNode, _, _)
    )
  }

  private module CustomPathGraphEdges {
    /**
     * An edge from the binding path in a view to a corresponding content of an internal model. e.g.
     * Given an XML declaration of a use of the control C, and a local model `oModel` it refers to:
     * ```xml
     * <C x="{/y}"/>
     * ```
     * and
     * ```javascript
     * let oModel = new JSONModel({ y: null });
     * ```
     *
     * establishes an edge from `x="{/y}"` to `y: null`.
     *
     * c.f. `FlowSteps::InternalModelContentToCustomMetadataPropertyStep`.
     */
    predicate bindingPathToInternalModelContent(
      UI5PathNode ui5PathNodePred, UI5PathNode ui5PathNodeSucc
    ) {
      exists(UI5BindingPath bindingPath |
        bindingPath = ui5PathNodePred.asUI5BindingPathNode() and
        ui5PathNodeSucc.asDataFlowNode() = bindingPath.getNode()
      )
    }

    /**
     * An edge from a content of an internal model to the corresponding binding path in a view,
     * which makes it an edge in the opposite direction as of `bindingPathToInternalModelContent`
     * above.
     * In order to ensure that the edge indeed holds, we check if the model's binding mode is
     * declared as two-way.
     *
     * c.f. `FlowSteps::InternalModelContentToCustomMetadataPropertyStep`.
     */
    predicate internalModelContentToBindingPath(
      UI5PathNode ui5PathNodePred, UI5PathNode ui5PathNodeSucc
    ) {
      exists(UI5BindingPath bindingPath, UI5InternalModel internalModel, Node boundNode |
        bindingPath = ui5PathNodeSucc.asUI5BindingPathNode() and
        boundNode = bindingPath.getNode() and
        (
          boundNode instanceof PropWrite and
          internalModel.(JsonModel).getAProperty() = boundNode // TODO: Generalize to UI5InternalModel
          or
          boundNode.asExpr() instanceof StringLiteral and
          ui5PathNodePred.asDataFlowNode() = boundNode
        ) and
        ui5PathNodePred.asDataFlowNode() = boundNode and
        internalModel.(JsonModel).isTwoWayBinding()
      )
    }
  }

  query predicate edges(UI5PathNode ui5PathNodePred, UI5PathNode ui5PathNodeSucc) {
    /* Include all existing dataflow edges */
    exists(ConfigPathNode pathNodeFrom, ConfigPathNode pathNodeTo |
      pathNodeFrom.getNode() = ui5PathNodePred.asDataFlowNode() and
      pathNodeTo.getNode() = ui5PathNodeSucc.asDataFlowNode() and
      ConfigPathGraph::edges(pathNodeFrom, pathNodeTo, _, _)
    ) and
    /* ========= TODO: Legacy code ========= */
    /* Exclude duplicate edge from model to handler parameter */
    not exists(UI5Handler h |
      ui5PathNodePred.asDataFlowNode() = h.getBindingPath().getNode() and
      ui5PathNodeSucc.asDataFlowNode() = h.getParameter(0)
    )
    or
    /* Flow to event handler parameter through the binding argument */
    exists(UI5Handler h |
      ui5PathNodePred.asUI5BindingPathNode() = h.getBindingPath() and
      ui5PathNodeSucc.asDataFlowNode() = h.getParameter(0)
    )
    or
    /* ===================================== */
    CustomPathGraphEdges::bindingPathToInternalModelContent(ui5PathNodePred, ui5PathNodeSucc)
    or
    CustomPathGraphEdges::internalModelContentToBindingPath(ui5PathNodePred, ui5PathNodeSucc)
    or
    /* Model with contents not statically observable */
    ui5PathNodePred.asUI5BindingPathNode().getModel() = ui5PathNodeSucc.asDataFlowNode() and
    ui5PathNodePred.asUI5BindingPathNode() = any(UI5View view).getASource()
    or
    ui5PathNodeSucc.asUI5BindingPathNode().getModel() = ui5PathNodePred.asDataFlowNode() and
    ui5PathNodeSucc.asUI5BindingPathNode() = any(UI5View view).getAnHtmlISink()
  }
}

module TrackPlaceAtCallConfig implements DataFlow::ConfigSig {
  /**
   * A child element instantiation.
   */
  predicate isSource(DataFlow::Node node) { node instanceof ElementInstantiation }

  /**
   * An "extension point" exposed from a parent element instantiation to
   * register a child to itself.
   */
  predicate isSink(DataFlow::Node node) { node instanceof ControlPlaceAtCall }

  /**
   * Step from data being written and the property that is being written to.
   * Needed to express the fact that the child control is usually written to
   * a property of a parent control to establish the child-parent relationship.
   *
   * NOTE: The "extension point" exposed by the parent control can come in the
   * form of:
   *
   * 1. `new Parent({ children: [ new Child( ... ) ] })`
   * 2. `new Parent(...).children[0] = new Child( ... )`
   * 3. `new Parent(...).addChild(new Child( ... ))`
   *
   * The first and second cases are (nested) `PropWrite`s to `getArgument(0)`
   * and are easily covered. But the third case poses a new problem of identifying
   * methods that writes to the `children` property, which CodeQL can't verify
   * because its definition is in the library.
   *
   * - [X] Mark all method calls: overgeneralizes, can infer dumb things
   *       (`sap.m.FlexBox.removeItem` comes to mind), but would still work for all
   *       true cases
   * - [ ] Heuristic by prefix: not an attractive option since heuristics can fail
   */
  predicate isAdditionalFlowStep(DataFlow::Node start, DataFlow::Node end) {
    exists(DataFlow::PropWrite propWrite |
      start = propWrite.getRhs() and
      end = propWrite.getBase()
    )
    or
    exists(DataFlow::MethodCallNode maybeAddingChildAPICall |
      start = maybeAddingChildAPICall.getAnArgument() and
      end = maybeAddingChildAPICall.getReceiver()
    )
    or
    exists(DataFlow::MethodCallNode call |
      start = call.getReceiver() and
      end = call
    )
    or
    exists(DataFlow::NewNode new |
      start = new.getAnArgument() and
      end = new
    )
  }
}
