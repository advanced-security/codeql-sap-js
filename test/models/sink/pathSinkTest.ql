/**
 * @id log-sinks
 * @name Path injection sinks
 * @kind problem
 */

import javascript
import models.UI5AMDModule
import semmle.javascript.security.dataflow.TaintedPathQuery

from Configuration config, DataFlow::Node sink
where config.isSink(sink, _)
select sink, sink.toString()
