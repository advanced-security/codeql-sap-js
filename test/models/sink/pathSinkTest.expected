| sink.js:53:54:53:58 | code1 | code1 |
| sink.js:55:56:55:60 | code1 | code1 |
| sink.js:57:36:57:40 | code0 | code0 |
| sink.js:59:42:59:46 | code0 | code0 |
| sink.js:61:46:61:50 | code0 | code0 |
| sink.js:65:39:65:43 | code0 | code0 |
| sink.js:67:38:67:42 | code0 | code0 |
| sink.js:69:33:69:37 | code0 | code0 |
| sink.js:71:37:71:41 | code0 | code0 |
| sink.js:73:36:73:40 | code0 | code0 |
| sink.js:73:43:73:47 | code1 | code1 |
| sink.js:75:37:75:41 | code0 | code0 |
| sink.js:75:44:75:48 | code1 | code1 |
| sink.js:77:40:77:44 | code0 | code0 |
| sink.js:77:47:77:51 | code1 | code1 |
| sink.js:79:40:79:44 | code0 | code0 |
| sink.js:79:47:79:51 | code1 | code1 |
| sink.js:94:28:94:32 | code0 | code0 |
| sink.js:96:49:96:53 | code1 | code1 |
| sink.js:98:28:98:32 | code0 | code0 |
